flocker/node/agents/testtools.py

```diff
@@ -80,14 +84,82 @@ def setUp(self):
     return Tests
 
 
-def cinder_client_from_environment():
+@implementer(INovaVolumeManager)
+class TidyNovaVolumeManager(
+        proxyForInterface(INovaVolumeManager, 'original')
+):
+    def __init__(self, original):
+        """
+        :param INovaVolumeManager original: An instance of
+            ``novaclient.v2.volumes.VolumeManager``.
+        """
+        self.original = original
+        self._attached_volumes = []
+
+    def create_server_volume(self, server_id, volume_id, device):
+        """
+        Wrap ``original.create_server_volume`` so as to record the
+        volumes that have been attached by this API so that they can
+        be later detached.
+        """
+        nova_attached_volume = self.original.create_server_volume(
+            server_id=server_id, 
+            volume_id=volume_id, 
+            device=device,
+        )
+        self._attached_volumes.append((server_id, nova_attached_volume))
+        return nova_attached_volume
+        
+    def _cleanup(self):
+        """
+        Detach any volumes that were attached by this API instance.
+        """
+        for server_id, volume in self._attached_volumes:
+            self.original.delete_server_volume(
+                server_id=server_id,
+                attachment_id=volume.id
+            )
+        # Don't return until all the volumes have been detached.
+        for server_id, nova_volume in self._attached_volumes:
+            wait_for_volume(
+                volume_manager=self.original,
+                expected_volume=nova_volume,
+                expected_status=u'available',
+            )
+            
+
+
+class INovaVolumeManagerTestsMixin(object):
+    """
+    Tests for ``INovaVolumeManager`` implementations.
+    """
+    def test_interface(self):
+        """
+        ``client`` provides ``INovaVolumeManager``.
+        """
+        self.assertTrue(verifyObject(INovaVolumeManager, self.client))
+
+
+def make_inovavolumemanager_tests(client_factory):
+    """
+    Build a ``TestCase`` for verifying that an implementation of
+    ``INovaVolumeManager`` adheres to that interface.
+    """
+    class Tests(INovaVolumeManagerTestsMixin, SynchronousTestCase):
+        def setUp(self):
+            self.client = client_factory(test_case=self)
+
+    return Tests
+
+
+def client_from_environment(client_factory):
     """
-    Create a ``cinderclient.v1.client.Client`` using credentials from a config
+    Create an openstack API client using credentials from a config
```

Comment:
It actually returns a client_factory, curried  with the   correct `session` and `region` arguments.
That way it can be used for creating both Nova and Cinder clients whose `version` arguments are different.